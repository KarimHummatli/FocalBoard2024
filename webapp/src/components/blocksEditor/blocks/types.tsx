import {ContentBlockTypes} from '../../../blocks/block'

export type BlockInputProps = {
    onChange: (value: string) => void
    value: string
    onCancel: () => void
    onSave: (val: string) => void
}

export type ContentType = {
    name: ContentBlockTypes
    displayName: string
    slashCommand: string
    prefix: string
    Input: React.FunctionComponent<BlockInputProps>
    render: (value: string) => React.ReactNode
    runSlashCommand: (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]) => void
    nextType?: string
}
