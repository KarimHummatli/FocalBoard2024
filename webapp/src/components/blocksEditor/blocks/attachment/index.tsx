import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './attachment.scss'

const Attachment: ContentType = {
    name: 'attachment',
    displayName: 'Attachment',
    slashCommand: '/attachment',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps) => <div className='AttachmentView'>ðŸ“Ž {props.value}</div>,
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Attachment'
                type='file'
                onChange={(e) => {
                    const files = e.currentTarget?.files
                    if (files) {
                        for (let i = 0; i < files.length; i++) {
                            const file = files.item(i)
                            if (file) {
                                props.onSave(file.name as string)
                            }
                        }
                    }
                }}
            />
        )
    }
}

Attachment.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Attachment)
    changeValue('')
}

export default Attachment
