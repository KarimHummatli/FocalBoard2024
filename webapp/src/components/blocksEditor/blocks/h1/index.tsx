import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h1.scss'

const H1: ContentType = {
    name: 'h1',
    displayName: 'Title',
    slashCommand: '/title',
    prefix: '# ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('# '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H1'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H1.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H1)
    changeValue(args.join(' '))
}

export default H1
