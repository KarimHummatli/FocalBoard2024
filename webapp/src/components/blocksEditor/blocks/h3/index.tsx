import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h3.scss'

const H3: ContentType = {
    name: 'h3',
    displayName: 'Sub Sub title',
    slashCommand: '/subsubtitle',
    prefix: '### ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('### '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H3'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H3.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H3)
    changeValue(args.join(' '))
}

export default H3
