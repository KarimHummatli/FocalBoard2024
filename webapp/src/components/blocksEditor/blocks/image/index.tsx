import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './image.scss'

type FileInfo = {
    file: string|ArrayBuffer
    width?: number
    align?: 'left'|'center'|'right'
}

const Image: ContentType<FileInfo> = {
    name: 'image',
    displayName: 'Image',
    slashCommand: '/image',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        if (props.value.file) {
            return <img className='ImageView' src={props.value.file as string}/>
        }
        return null
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <div>
                {props.value.file && (typeof props.value.file === 'string') && (
                    <img
                        className='ImageView' src={props.value.file}
                        onClick={() => ref.current?.click()}
                    />
                )}
                <input
                    ref={ref}
                    className='Image'
                    type='file'
                    accept='image/*'
                    onChange={(e) => {
                        const reader = new FileReader();
                        const file = (e.currentTarget?.files || [])[0]
                        reader.readAsArrayBuffer(file);
                        reader.onload = function () {
                            props.onSave({file: reader.result as ArrayBuffer})
                        }
                    }}
                />
            </div>
        )
    }
}

Image.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Image)
    changeValue({file: ''})
}

export default Image
