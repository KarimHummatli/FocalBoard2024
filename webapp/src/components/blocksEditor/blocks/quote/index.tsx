import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './quote.scss'

const Quote: ContentType = {
    name: 'quote',
    displayName: 'Quote',
    slashCommand: '/quote',
    prefix: '> ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('> '+value, {renderer, breaks: true})
        return <div
            className='Quote'
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <blockquote className='Quote'>
                <input
                    ref={ref}
                    onChange={(e) => props.onChange(e.currentTarget.value)}
                    onKeyDown={(e) => {
                        if (props.value === '' && e.key === "Backspace") {
                            props.onCancel()
                        }
                        if (e.key === "Enter") {
                            props.onSave(props.value)
                        }
                    }}
                    onBlur={() => props.onSave(props.value)}
                    value={props.value}
                />
            </blockquote>
        )
    }
}

Quote.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(Quote)
    changeValue(args.join(' '))
}

export default Quote
