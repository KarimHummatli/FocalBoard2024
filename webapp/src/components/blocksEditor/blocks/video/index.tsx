import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './video.scss'

const Video: ContentType = {
    name: 'video',
    displayName: 'Video',
    slashCommand: '/video',
    prefix: '',
    render: (value: string) => (
        <video width="320" height="240" controls className='VideoView'>
            <source src={value} type='video/mp4'/>
        </video>
    ),
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Video'
                type='file'
                accept='video/*'
                onChange={(e) => {
                    const reader = new FileReader();
                    const file = (e.currentTarget?.files || [])[0]
                    reader.readAsDataURL(file);
                    reader.onload = function () {
                        props.onSave(reader.result as string)
                    }
                }}
            />
        )
    }
}

Video.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Video)
    changeValue('')
}

export default Video
