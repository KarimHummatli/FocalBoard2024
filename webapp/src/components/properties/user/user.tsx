// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useContext} from 'react'
import Select from 'react-select'

import {IUser, WorkspaceUsersContext, WorkspaceUsersContextData} from '../../../user'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
    tooltip?: string
}

const UserProperty = (props: Props): JSX.Element => {
    const workspaceUsers = useContext<WorkspaceUsersContextData>(WorkspaceUsersContext)
    const tooltipClassName = props.tooltip ? 'octo-tooltip tooltip-top' : ''

    if (props.readonly) {
        const userClassName = `UserProperty octo-propertyvalue ${tooltipClassName}`
        return (
            <div
                className={userClassName}
                data-tooltip={props.tooltip}
            >
                {workspaceUsers.usersById.get(props.value)?.username || props.value}</div>)
    }

    return (
        <Select
            options={workspaceUsers.users}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={'UserProperty'}
            styles={getSelectBaseStyle()}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={workspaceUsers.usersById.get(props.value) || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
