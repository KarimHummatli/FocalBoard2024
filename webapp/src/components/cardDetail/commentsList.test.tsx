// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'
import 'isomorphic-fetch'

import {render} from '@testing-library/react'

import {act} from 'react-dom/test-utils'

import {CommentBlock} from '../../blocks/commentBlock'

import {wrapIntl} from '../../testUtils'

import {FetchMock} from '../../test/fetchMock'
import {IUser} from '../../user'

import CommentsList from './commentsList'

global.fetch = FetchMock.fn

beforeEach(() => {
    FetchMock.fn.mockReset()
})

describe('components/cardDetail/CommentsList', () => {
    const createdAt = Date.parse('01 Jan 2021 00:00:00 GMT')
    const comment1: CommentBlock = {
        id: 'comment_id_1',
        title: 'Comment 1',
        createAt: createdAt,
        modifiedBy: 'user_id_1',
    } as CommentBlock

    const comment2: CommentBlock = {
        id: 'comment_id_2',
        title: 'Comment 2',
        createAt: createdAt,
        modifiedBy: 'user_id_2',
    } as CommentBlock

    test('comments show up', async () => {
        FetchMock.fn.mockReturnValue(FetchMock.jsonResponse(JSON.stringify({username: 'username_1'} as IUser)))
        const component = wrapIntl(
            <CommentsList
                comments={[comment1, comment2]}
                rootId={'root_id'}
                cardId={'card_id'}
                readonly={false}
            />,
        )

        let container

        await act(async () => {
            const result = render(component)
            container = result.container
        })

        expect(container).toMatchSnapshot()
    })

    test('comments show up in readonly mode', async () => {
        FetchMock.fn.mockReturnValue(FetchMock.jsonResponse(JSON.stringify({username: 'username_1'} as IUser)))
        const component = wrapIntl(
            <CommentsList
                comments={[comment1, comment2]}
                rootId={'root_id'}
                cardId={'card_id'}
                readonly={true}
            />,
        )

        let container

        await act(async () => {
            const result = render(component)
            container = result.container
        })

        expect(container).toMatchSnapshot()
    })
})
