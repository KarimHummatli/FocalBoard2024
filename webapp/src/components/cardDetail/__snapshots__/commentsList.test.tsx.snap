// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/cardDetail/CommentsList comments show up 1`] = `
<div>
  <div
    class="CommentsList"
  >
    <div
      class="Comment comment"
    >
      <div
        class="comment-header"
      >
        <img
          class="comment-avatar"
          src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
        />
        <div
          class="comment-username"
        >
          username_1
        </div>
        <div
          class="comment-date"
        >
          January 01, 5:30 AM
        </div>
        <div
          class="MenuWrapper"
        >
          <div
            class="Button IconButton"
          >
            <i
              class="CompassIcon icon-dots-horizontal OptionsIcon"
            />
          </div>
        </div>
      </div>
      <div
        class="comment-text"
      >
        <p>
          Comment 1
        </p>
      </div>
    </div>
    <div
      class="Comment comment"
    >
      <div
        class="comment-header"
      >
        <img
          class="comment-avatar"
          src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
        />
        <div
          class="comment-username"
        />
        <div
          class="comment-date"
        >
          January 01, 5:30 AM
        </div>
        <div
          class="MenuWrapper"
        >
          <div
            class="Button IconButton"
          >
            <i
              class="CompassIcon icon-dots-horizontal OptionsIcon"
            />
          </div>
        </div>
      </div>
      <div
        class="comment-text"
      >
        <p>
          Comment 2
        </p>
      </div>
    </div>
    <div
      class="commentrow"
    >
      <img
        class="comment-avatar"
        src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
      />
      <div
        class="MarkdownEditor octo-editor newcomment "
      >
        <div
          class="octo-editor-preview octo-placeholder"
        >
          <p>
            Add a comment...
          </p>
        </div>
        <div
          class="octo-editor-activeEditor"
          style="visibility: hidden; position: absolute; top: 0px; left: 0px;"
        >
          <div
            id="simplemde-editor-1-wrapper"
          >
            <textarea
              id="simplemde-editor-1"
              style="display: none;"
            />
            <div
              class="EasyMDEContainer"
            >
              <div
                class="CodeMirror cm-s-easymde CodeMirror-wrap"
              >
                <div
                  style="overflow: hidden; position: relative; width: 3px; height: 0px;"
                >
                  <textarea
                    autocapitalize="off"
                    autocorrect="off"
                    spellcheck="false"
                    style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                    tabindex="0"
                  />
                </div>
                <div
                  class="CodeMirror-vscrollbar"
                  cm-not-content="true"
                  tabindex="-1"
                >
                  <div
                    style="min-width: 1px;"
                  />
                </div>
                <div
                  class="CodeMirror-hscrollbar"
                  cm-not-content="true"
                  tabindex="-1"
                >
                  <div
                    style="height: 100%; min-height: 1px;"
                  />
                </div>
                <div
                  class="CodeMirror-scrollbar-filler"
                  cm-not-content="true"
                />
                <div
                  class="CodeMirror-gutter-filler"
                  cm-not-content="true"
                />
                <div
                  class="CodeMirror-scroll"
                  style="min-height: 10px;"
                  tabindex="-1"
                >
                  <div
                    class="CodeMirror-sizer"
                    style="margin-left: 0px;"
                  >
                    <div
                      style="position: relative;"
                    >
                      <div
                        class="CodeMirror-lines"
                        role="presentation"
                      >
                        <div
                          role="presentation"
                          style="position: relative; outline: none;"
                        >
                          <div
                            class="CodeMirror-measure"
                          >
                            <pre
                              class="CodeMirror-line-like"
                            >
                              <span>
                                xxxxxxxxxx
                              </span>
                            </pre>
                          </div>
                          <div
                            class="CodeMirror-measure"
                          />
                          <div
                            style="position: relative; z-index: 1;"
                          />
                          <div
                            class="CodeMirror-cursors"
                          />
                          <div
                            class="CodeMirror-code"
                            role="presentation"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    style="position: absolute; height: 50px; width: 1px;"
                  />
                  <div
                    class="CodeMirror-gutters"
                    style="display: none;"
                  />
                </div>
              </div>
              <div
                class="editor-preview-side editor-preview"
              />
            </div>
          </div>
        </div>
      </div>
      
    </div>
    <hr />
  </div>
</div>
`;

exports[`components/cardDetail/CommentsList comments show up in readonly mode 1`] = `
<div>
  <div
    class="CommentsList"
  >
    <div
      class="Comment comment"
    >
      <div
        class="comment-header"
      >
        <img
          class="comment-avatar"
          src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
        />
        <div
          class="comment-username"
        >
          username_1
        </div>
        <div
          class="comment-date"
        >
          January 01, 5:30 AM
        </div>
      </div>
      <div
        class="comment-text"
      >
        <p>
          Comment 1
        </p>
      </div>
    </div>
    <div
      class="Comment comment"
    >
      <div
        class="comment-header"
      >
        <img
          class="comment-avatar"
          src="data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 100 100\\" style=\\"fill: rgb(192, 192, 192);\\"><rect width=\\"100\\" height=\\"100\\" /></svg>"
        />
        <div
          class="comment-username"
        />
        <div
          class="comment-date"
        >
          January 01, 5:30 AM
        </div>
      </div>
      <div
        class="comment-text"
      >
        <p>
          Comment 2
        </p>
      </div>
    </div>
    <hr />
  </div>
</div>
`;
